hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
+                value: 0,
+                span: Span::call_site(),
+            };
+            // A -> H.
+            new_succs[src].insert(*src_idx, (hoff_id, zero_index));
+            // A <- H.
+            new_preds[hoff_id].insert(zero_index, (src, *src_idx));
+            // H <- Z.
+            new_preds[dst].insert(*dst_idx, (hoff_id, zero_index));
+            // H -> Z.
+            new_succs[hoff_id].insert(zero_index, (dst, *dst_idx));
+        }
+    }
 
-            // Already has a handoff, no need to insert one.
-            if is_subgraph_edge
-                || matches!(flat_graph.nodes[src], Node::Handoff)
-                || matches!(flat_graph.nodes[dst], Node::Handoff)
-            {
-                new_preds[dst].insert(*dst_idx, (src, *src_idx));
-                new_succs[src].insert(*src_idx, (dst, *dst_idx));
-            } else {
-                // Needs handoff inserted.
-                // A -> H -> Z
-                let hoff_id = flat_graph.nodes.insert(Node::Handoff);
-                new_preds.insert(hoff_id, Default::default());
-                new_succs.insert(hoff_id, Default::default());
-
-                let zero_index = IndexInt {
-                    value: 0,
-                    span: Span::call_site(),
-                };
-                // A -> H.
-                new_succs[src].insert(*src_idx, (hoff_id, zero_index));
-                // A <- H.
-                new_preds[hoff_id].insert(zero_index, (src, *src_idx));
-                // H <- Z.
-                new_preds[dst].insert(*dst_idx, (hoff_id, zero_index));
-                // H -> Z.
-                new_succs[hoff_id].insert(zero_index, (dst, *dst_idx));
+    // Determine node's subgraph and subgraph's nodes.
+    // This list of nodes in each subgraph are to be in topological sort order.
+    let (node_subgraph, subgraph_nodes) = {
+        // Ignore handoffs. Esentially each subgraph has it's own topo sort.
+        let topo_sort = graph_algorithms::topo_sort(
+            nodes
+                .iter()
+                .filter(|&(_, node)| !matches!(node, Node::Handoff))
+                .map(|(node_id, _)| node_id),
+            |v| {
+                new_preds
+                    .get(v)
+                    .into_iter()
+                    .flatten()
+                    .map(|(_src_idx, &(dst, _dst_idx))| dst)
+                    .filter(|&dst| !matches!(nodes[dst], Node::Handoff))
+            },
+        );
+
+        // For a `NodeId`, what `SubgraphId` does it belong to.
+        let mut node_subgraph: SecondaryMap<GraphNodeId, GraphSubgraphId> = Default::default();
+        // For a `SubgraphId`, what `NodeId`s belong to it.
+        let mut subgraph_nodes: SlotMap<GraphSubgraphId, Vec<GraphNodeId>> = Default::default();
+        // Populate above.
+        let mut grouped_nodes: SecondaryMap<GraphNodeId, Vec<GraphNodeId>> = Default::default();
+        for node_id in topo_sort {
```

Comment:
do we have to iterate through `node_id`s? Could we iterate through the disjoint sets from UnionFind and bulk-handle some things? Would that be more efficient?

Reply:
keeping as-is